runtime/interpreter/interpreter.go

```diff
@@ -2150,46 +2150,52 @@ func (interpreter *Interpreter) convert(value Value, valueType, targetType sema.
 		}
 
 	case *sema.ReferenceType:
-		if !valueType.Equal(unwrappedTargetType) {
-			// transferring a reference at runtime does not change its entitlements; this is so that an upcast reference
-			// can later be downcast back to its original entitlement set
-
-			// check defensively that we never create a runtime mapped entitlement value
-			if _, isMappedAuth := unwrappedTargetType.Authorization.(*sema.EntitlementMapAccess); isMappedAuth {
-				panic(UnexpectedMappedEntitlementError{
-					Type:          unwrappedTargetType,
-					LocationRange: locationRange,
-				})
-			}
+		targetAuthorization := ConvertSemaAccessToStaticAuthorization(interpreter, unwrappedTargetType.Authorization)
+		switch ref := value.(type) {
+		case *EphemeralReferenceValue:
+			if !(ref.BorrowedType.Equal(unwrappedTargetType.Type) && ref.Authorization.Equal(targetAuthorization)) {
```

Comment:
Instead of `valueType` type (`ReferenceStaticType`) which is always the same as the type of the referenced value, use the "borrow type" of the value to determine whether this value should be converted to the target type or not.

Reply:
Should we also not convert if the `valueType` is different, even if the `borrowedType` is the same? E.g. if we have two references that both point to an underlying `R` value, but one has an `AnyResource` borrow type and the other one is `R`? Why don't we need to convert in that case?

Reply:
That's the thing, `valueType` is no longer the borrow type. SO in your example, even if the borrow types are different, `valueType` remains the same.

We can still keep the old check just to be safe.

Reply:
Yeah, I think we should convert if any of {`BorrowedType`, `Authorization`, or `valueType`} are different